subsys/caf/modules/sensor_sampler.c

```diff
@@ -134,35 +152,46 @@ static bool is_active(const struct sensor_config *sc, struct sensor_data *sd,
 	return is_active;
 }
 
-static bool can_sensor_sleep(const struct sensor_config *sc,
-			     struct sensor_data *sd,
-			     const float *curr)
+static void reset_sensor_trigger_counter(const struct sensor_config *sc,
+					 struct sensor_data *sd)
 {
-	size_t data_cnt = get_sensor_data_cnt(sc);
-	bool sleep = true;
+	unsigned int timeout = sc->trigger ? sc->trigger->activation.timeout_ms : 0;
+	unsigned int period = sc->sampling_period_ms;
 
-	for (size_t i = 0; i < data_cnt; i++) {
-		if (is_active(sc, sd, curr[i], sd->prev[i])) {
-			sleep = false;
-			break;
-		}
-	}
+	sd->sleep_cntd = ceiling_fraction(timeout, period);
+}
 
-	if (sleep) {
-		unsigned int timeout = sc->trigger->activation.timeout_ms;
-		unsigned int period = sc->sampling_period_ms;
+static void process_sensor_activity(const struct sensor_config *sc,
+				    struct sensor_data *sd,
+				    const float *curr)
+{
+	if (sc->trigger) {
+		size_t data_cnt = get_sensor_data_cnt(sc);
+		bool sleep = true;
 
-		sd->sleep_cnt++;
-		if (sd->sleep_cnt >= timeout/period) {
-			sd->sleep_cnt = 0;
-			return true;
+		for (size_t i = 0; i < data_cnt; i++) {
+			if (trigger_check_active(sc, sd, curr[i], sd->prev[i])) {
+				sleep = false;
+				break;
+			}
+		}
+
+		if (sleep) {
+			if (sd->sleep_cntd > 0) {
+				--(sd->sleep_cntd);
```

Comment:
btw. What's the advantage of this instead of having this as it used to be? I don't get why you rewrote this to count down instead of counting up. I agree there should be `ceiling_fraction` but was it really necessary to rewrite the logic?